### ISSUE TITLE
lost context when use @fastify/multipart

### ISSUE THREAD

**Issue by wvq** [fastify/fastify-request-context#153]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.15.0

### Plugin version

4.2.0

### Node.js version

18.14.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

11.3

### Description

when register `multipart` after `context`,  the router frequently lost context info

### Steps to Reproduce

```javascript
const Fastify = require('fastify').default
const MultipartSupport = require('@fastify/multipart')
const { requestContext } = require('@fastify/request-context')
const RequestContextSupport = require('@fastify/request-context')

const server = Fastify({ logger: true })


server.register(RequestContextSupport, {
  defaultStoreValues: {
    user: { id: 0 },
  },
})

server.addHook('onRequest', async (req, reply) => {
  req.requestContext.set('user', { id: 1 })
})

server.register(MultipartSupport, {
  addToBody: true,
})

server.post('/', async (req, reply) => {
  console.log(req.requestContext.get('user'))

  return { message: 'ok' }
})

server.listen({ host: '0.0.0.0', port: 3002 })

```
use `postman` to send any file to `http://127.0.0.1/`, 
`console.log(req.requestContext.get('user'))` sometimes are `undefined`

**Comment by climba03003** [comment]
The comment from  source code already explain why it happen.
It is the limitation of `EventEmitter` async resources.

https://github.com/fastify/fastify-request-context/blob/e5a722071c104b6785eced39e23e14176f38511e/index.js#L32-L41

It would be better to point it out in the readme.
Would you like to work on this?

**Comment by kibertoad** [comment]
@climba03003 Shouldn't the commented block help restore the correct context in that case? It was added to fix the context being lost when using body-parser, I wonder if some additional tweak is needed there to correctly handle multipart parsing hook as well.

**Comment by climba03003** [comment]
> Shouldn't the commented block help restore the correct context in that case?

Yes, it also explain why the parser will mess up the context because `busboy` also an EventEmitter.

**Comment by wvq** [comment]
@climba03003 @kibertoad 
Thanks for reply.  But it still confused me, as it run on different async context, why sometimes correct output context?

**Comment by kibertoad** [comment]
I'll write a couple of tests, but I think that it should work correctly if you register request-context plugin after the multipart one, so that it gets a chance to rebind the context correctly

**Comment by wvq** [comment]
Yes, I register `request-context` after `multipart` to avoid this problem, it works fine.

In my situaction, I want:
> 1. get header properties then set `user info` to request-context.
> 2. in multipart `onFile` callback, store files in difference directory base on the `user info` 
> 3. process other logic in some routes also need `user info` (has lost context problem)

Now, Call `step 2` first,  means the `onFile` callback need resolve `step 1` logic;
 and then call `step 1`  to set context, it a bad performance.

To avoid performance issue, I need to set property to `req`,  and `step 1` check if the property exists, the code is ugly.

**Comment by mcollina** [comment]
I think we should solve this in our multipart module by adding a AsyncResource at the right moment.

**Comment by mcollina** [comment]
Thanks for reporting! Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by gustakasn0v** [comment]
@climba03003 @mcollina Now that https://github.com/fastify/fastify-request-context/pull/173 has been merged, does that mean this issue can be resolved? Or is there still work left to fix compatibility with @fastify/multipart?

Asking so I can at least raise a quick PR to add this gotcha to the readme

**Comment by Alt-er** [comment]
In the @fastify/request-context plugin implementation, I noticed this pattern:
```
asyncLocalStorage.run({ ...defaultStoreValues }, () => {
      const asyncResource =
        opts.createAsyncResource != null
          ? opts.createAsyncResource(req, requestContext)
          : new AsyncResource('fastify-request-context')
      req[asyncResourceSymbol] = asyncResource
      asyncResource.runInAsyncScope(done, req.raw)  // <-- Why not just done()?
    })
```
What's the practical difference between using asyncResource.runInAsyncScope(done, req.raw) and calling done() directly in this context?
